### DISCUSSION TITLE
Expected path System.Collections.Hashtable to exist, but it did not exist.

### DISCUSSION THREAD

**Discussion by riosengineer** [pester/Pester#2422] [Q&A] [answered]
Hey all, 

Looking for some guidance or a solution. I'm a bit stuck. My test seems to work fine but it seems to give this error even though my tests pass. I'm using Pester v5 in an Azure DevOps pipeline with a PowerShell task. 

I'm essentially collecting all files in a specific path, then joining these to create an array with a list of file paths to run a 'Test-Path' / Should -Exist test. 

```
Starting discovery in 1 files.
Discovery found 25 tests in 326ms.
Running tests.
##[error] [-] .tests file. should exist 110ms (86ms|24ms)
##[error]  Expected path System.Collections.Hashtable to exist, but it did not exist.
##[error]  at $_ | Should -Exist, /home/vsts/work/1/s/pipelines/.tests/pester/tests.ps1:17
##[error]  at <ScriptBlock>, /home/vsts/work/1/s/pipelines/.tests/pester/tests.ps1:17
Tests completed in 1.19s
Tests Passed: 24, Failed: 1, Skipped: 0 NotRun: 0

##[error]PowerShell exited with code '1'.
```

My test: 

```
BeforeDiscovery {
    # Change the current directory to the Bicep module folder
    Set-Location -Path "/modules"
    # Define the module and test paths
    $testPath = "/home/vsts/work/1/s/modules/.tests"
    $moduleFiles = Get-ChildItem -Filter *.bicep

    foreach ($moduleFile in $moduleFiles) {
    $testFile = Join-Path $testPath ($moduleFile.BaseName + ".tests.bicep")
    $testFiles = @($testFiles) + @($testFile)
    }
}

Describe "test file" -ForEach $testFiles {
    # Check if a corresponding .tests.bicep file exists for each .bicep module file
    It "$_ should exist" {
        $_ | Should -Exist 
    }
}
```

It is running the tests successfully. It's like it's adding an extra `Should -Exist` that I can't understand currently. If I print the $testFiles out to Write-Host it all looks as expected. 

What am I missing here? 

Thank you!

**Comment by fflaten** [comment, marked as answer]
Hi. `$testFiles = @($testFiles) + @($testFile)` adds a null-element at the start of your array due to  `$testFiles` being undefined on the first foreach-iteration. The test generated for the `$null`-element has `$_` set to an empty hashtable atm. Related #2320.

To avoid the null-element, you can either:
- Declare `$testFiles = @()` before the loop
- Or avoid the array merging in the first place. It is inefficient anyways. Try:
```powershell
    $testFiles = foreach ($moduleFile in $moduleFiles) {
        Join-Path $testPath ($moduleFile.BaseName + '.tests.bicep')
    }
```

**Reply by riosengineer** [reply]
Hey @fflaten. Thank you for this - your efficiency fix worked a treat and all is well. Moreover, I learnt something. I appreciate you taking the time to help.
